ref(generic-metrics): Migrate UseCaseKey to UseCaseID for organization metrics

john-z-yang opened the pull request:
### Refactoring usages of `UseCaseKey` into `UseCaseID`

After recent changes to the metrics platform, we are no longer using `UseCaseKey` to represent the use case of a metrics and are transitioning to `UseCaseID`.

This pr refactors instances of `UseCaseKey` so that in the future it can be deleted/deprecated (see pr [46575](https://github.com/getsentry/sentry/pull/46575))

iambriccardo commented on src/sentry/snuba/metrics/fields/base.py:
```diff
@@ -195,11 +196,11 @@ def _get_entity_of_metric_mri(
 
     entity_keys_set: frozenset[EntityKey]
-    if use_case_id == UseCaseKey.PERFORMANCE:
+    if use_case_id == UseCaseID.TRANSACTIONS:
```
```suggestion
    if use_case_id in [UseCaseID.TRANSACTIONS, UseCaseID.SPANS]:
```

Assuming `generic_metrics` are also used for spans.

iambriccardo replied:
```suggestion
    METRIC_PATH_MAPPING[use_case_id] is UseCaseKey.PERFORMANCE
```

Or like you did in the other change but I would prefer to just be explicit about it and use the `in []`. What do you think?

john-z-yang replied:
Thanks! Aa discussed offline, I added a function in use_case_id_registry.py to retrieve the UseCaseKey

iambriccardo commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -378,8 +378,8 @@ def validate_granularity(self) -> None:
-            if self.use_case_key == UseCaseKey.RELEASE_HEALTH or (
-                self.use_case_key == UseCaseKey.PERFORMANCE and not self.include_series
+            if self.use_case_id == UseCaseID.SESSIONS or (
+                self.use_case_id == UseCaseID.TRANSACTIONS and not self.include_series
```
```suggestion
                self.use_case_id in [UseCaseID.TRANSACTIONS, UseCaseID.SPANS] and not self.include_series
```

iambriccardo commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -400,7 +400,7 @@ def __post_init__(self) -> None:
 
         if (
-            self.use_case_key == UseCaseKey.PERFORMANCE
+            self.use_case_id == UseCaseID.TRANSACTIONS
```
```suggestion
            self.use_case_id in [UseCaseID.TRANSACTIONS, UseCaseID.SPANS]
```

iambriccardo commented on src/sentry/snuba/metrics/query_builder.py:
```diff
@@ -951,7 +952,7 @@ def __build_totals_and_series_queries(
             series_limit = limit.limit * intervals_len
 
-            if self._use_case_id == UseCaseKey.PERFORMANCE:
+            if METRIC_PATH_MAPPING[self._use_case_id] is UseCaseKey.PERFORMANCE:
```
```suggestion
            if self._use_case_id in [UseCaseID.TRANSACTIONS, UseCaseID.SPANS]
```

iambriccardo commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -166,19 +166,19 @@ def projects(self) -> QuerySet:
         elif parsed_mri.namespace == "sessions":
-            return UseCaseKey.RELEASE_HEALTH
+            return UseCaseID.SESSIONS
         raise ValueError("Can't find correct use_case_id based on metric MRI")
```
You could also extend this function to support `spans`.

iambriccardo reviewed (commented):
Great work! I have left a few comments that I think should be addressed iff we know that `UseCaseID.SPANS` will use the `generic_metrics` dataset. If not then just remove any check for `SPANS` in the code (limiting it to only `TRANSACTIONS` and `SESSIONS`), since we might have to support spans for the metrics layer in the future.

john-z-yang commented:
> Great work! I have left a few comments that I think should be addressed iff we know that `UseCaseID.SPANS` will use the `generic_metrics` dataset. If not then just remove any check for `SPANS` in the code (limiting it to only `TRANSACTIONS` and `SESSIONS`), since we might have to support spans for the metrics layer in the future.

Thanks, addressed the comments. I also had a chat with the rest of my team and can confirmed that `UseCaseKey` is guaranteed to reflect the dataset

iambriccardo commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -341,7 +340,7 @@ def validate_end(self) -> None:
         # Logic specific to how we handle time series in discover in terms of granularity and interval
         if (
-            self.use_case_key == UseCaseKey.PERFORMANCE
+            self.use_case_id == UseCaseID.TRANSACTIONS
```
Why are you not using `get_use_case_key` also here? To be safe and consistent, I would prefer to use it across the layer in order to keep it logically similar to before (ofc now spans will be supported out of the box but I am not sure they would work right away).

john-z-yang replied:
Sure, I was hesitant to use `get_use_case_key` here because like you said, I was hesitant whether or not span will work right away.

RaduW replied:
@iambriccardo I'm not sure what are you suggesting ( at this level we shouldn't see a `UseCaseKey` anymore, everything should be UseCaseID, if we don't support spans here we should make it explicit).

iambriccardo replied:
I changed my mind around it, at first I was like, lets convert every incoming `use_case_id` to `use_case_key` in order to keep the logic identical but that will imply that `SPANS` will be implicitly supported, which is not good.

Sorry for the confusion.

iambriccardo reviewed (commented):
The changes are fine but I don't get why in some places you used `get_use_case_key` and in some directly the use case id. I think that for uniformity we should agree to either convert everything in the layer to use case id or perform the transformations. I am just suggesting this, since if we do coercions in some places and in others we don't, the very deep call stack of a metrics layer query might become very difficult to debug.

iambriccardo commented:
I am starting to think that the best way to do this would be to convert `PERFORMANCE` to `TRANSACTIONS` in code and whenever my team will add spans support, we will take care of the coercions. How does that sound to you?

john-z-yang commented:
> I am starting to think that the best way to do this would be to convert `PERFORMANCE` to `TRANSACTIONS` in code and whenever my team will add spans support, we will take care of the coercions. How does that sound to you?

Do you mean not use the `get_use_case_key ` and just compare by `== UseCaseKey.` for now?

iambriccardo commented:
> > I am starting to think that the best way to do this would be to convert `PERFORMANCE` to `TRANSACTIONS` in code and whenever my team will add spans support, we will take care of the coercions. How does that sound to you?
> 
> Do you mean not use the `get_use_case_key` and just compare by `== UseCaseKey.` for now?

I meant just placing `UseCaseID.TRANSACTIONS` everywhere where `UseCaseKey.PERFORMANCE` was and then me and my team would effectively use the utils function that you wrote to wire in spans iff we would need to.

iambriccardo reviewed (approved):
LGTM, will wait for @RaduW's double check here.

RaduW commented on src/sentry/snuba/metrics/query.py:
```diff
@@ -378,8 +376,8 @@ def validate_granularity(self) -> None:
-            if self.use_case_key == UseCaseKey.RELEASE_HEALTH or (
-                self.use_case_key == UseCaseKey.PERFORMANCE and not self.include_series
+            if self.use_case_id is UseCaseID.SESSIONS or (
+                self.use_case_id is UseCaseID.TRANSACTIONS and not self.include_series
```
```suggestion
                self.use_case_id in {UseCaseID.TRANSACTIONS, UseCaseID.SPANS} and not self.include_series
```

john-z-yang replied:
It was suggested by @iambriccardo that we should [hold off on any changes regarding spans for now](https://github.com/getsentry/sentry/pull/51776#issuecomment-1617538339). And the telemetry team will add it later.

RaduW commented on tests/sentry/api/endpoints/test_organization_metric_data.py:
```diff
@@ -70,7 +70,7 @@ def test_incorrect_use_case_id_value(self):
         assert (
             response.json()["detail"]
-            == "Invalid useCase parameter. Please use one of: ['release-health', 'performance']"
+            == f"Invalid useCase parameter. Please use one of: {[uc.value for uc in UseCaseID]}"
```
(as commented in the endpoint)

I don't think we should change the inputs of endpoints at this time.
If we decide to do it 
* we should first accept both UseCaseKey and UseCaseID in the endpoints
* change the UI (second PR)
* change the endpoints to accept only UseCaseID ( 3rd PR )

RaduW reviewed (changes requested):
Really nice PR, thank you for doing it.

Let's clarify the endpoint issue (below) first and then I'll approve it.

One big concern I have is regarding changing the input/output of endpoints from "performance",  "release-health" to "transactions" and "sessions".

I think we can't do that in one go.
If we want to do it we would have to: 

* first support both as input (in the endpoints) 
* change UI to send UseCaseIDs and accept both UseCaseIDs and UseCaseKeys
* change endpoints to only accept UseCaseIDs and send UseCaseIDs
* change UI to accept only UseCaseIDs

One small concern: I think we should not accept both UseCaseID and UseCaseKey in the `sentry_metrics.utils`. 

I think you already handled all cases already so there is no need to keep the mess around.

john-z-yang commented on src/sentry/sentry_metrics/use_case_id_registry.py:
```diff
@@ -40,5 +40,5 @@ class UseCaseID(Enum):
 
-def get_metric_path_from_usecase(use_case: UseCaseID) -> UseCaseKey:
-    return METRIC_PATH_MAPPING[use_case]
+def get_use_case_key(use_case_id: UseCaseID) -> Optional[UseCaseKey]:
+    return METRIC_PATH_MAPPING.get(use_case_id)
```
`UseCaseIDs` will always map to `UseCaseKey`, the Optional is from a conversation offline with @iambriccardo where we want this function to never throw. In any case, this is not being used in right now and is left as a utility to be used in the future.

john-z-yang commented on src/sentry/api/endpoints/organization_metrics.py:
```diff
@@ -21,17 +21,17 @@
 
     try:
-        return UseCaseKey(request.GET.get("useCase", "release-health"))
+        return UseCaseID(request.GET.get("useCase", "sessions"))
     except ValueError:
         raise ParseError(
-            detail=f"Invalid useCase parameter. Please use one of: {[uc.value for uc in UseCaseKey]}"
+            detail=f"Invalid useCase parameter. Please use one of: {[uc.value for uc in UseCaseID]}"
         )
```
Do you mean our API should accept the following for now
| strings it accept | enum it maps to |
|----|---|
|`"performance"` | `UseCaseID.TRANSACTIONS`|
|`"transactions"` | `UseCaseID.TRANSACTIONS`|
|`"sessions"`| `UseCaseID.SESSIONS`|
|`"release-health"`| `UseCaseID.SESSIONS`|


Once this pr is deployed, and the front end changes made, we will change the API again so that it will accept:


| strings it accept | enum it maps to |
|----|---|
|`"transactions"` | `UseCaseID.TRANSACTIONS`|
|`"sessions"`| `UseCaseID.SESSIONS`|

RaduW replied:
Hey John,

If it was up to me I would just keep the external interface fixed ( i.e. emit and accept the strings "performance" and "release-health" . 

What you propose is much cleaner, and also harder.

If we change our endpoint to also accept "transactions" and "sessions" there is no harm done, we are still backward compatible.

At that point in time we can also change the UI to both accept and send "transactions" and "sessions".

Changing the backend to emit (in the response) "transactions" and "sessions" or stop accepting "release-health" and "performance" in its parameters breaks backward compatibility and I'm not quite sure what is our policy on that. 

**We need to take in account that we potentially have external users that are using the API (all our endpoints are considered public API) and if we do that we'll break them ( for no good reason IMO).**

Before breaking compatibility we should check if that's acceptable (not sure who we should ask).

As a compromise we could make the Backend accept "transactions" and "sessions" ( in addition to "release-health" and "performance" ) and maybe make the UI send "transactions" and "sessions" and also accept "transactions" and "sessions" in addition to "release-health" and "performance".

john-z-yang replied:
I see, thanks!

john-z-yang commented:
@RaduW and @iambriccardo do you guys mind giving me a hand with this test single test that is failing?
```
tests/sentry/api/endpoints/test_organization_metric_tags.py::OrganizationMetricsTagsIntegrationTest::test_mri_metric_tags__in_region_silo - AssertionError: (500, b'{"detail":"Internal Error","errorId":null}')
```
I think I might have missed something really obvious but can't figure out why :(

RaduW commented:
@john-z-yang 
`_fetch_tags_or_values_for_mri` had a parameter changed from `use_case_key` to `use_case_id` but not all calls were refactored ( see commit above).

obostjancic commented on src/sentry/testutils/cases.py:
```diff
@@ -3,7 +3,7 @@
 import responses
 import sentry_kafka_schemas
 
-from sentry.sentry_metrics.use_case_id_registry import REVERSE_METRIC_PATH_MAPPING, UseCaseID
```
Can `REVERSE_METRIC_PATH_MAPPING` be removed completely? According to [this comment](https://github.com/getsentry/sentry/blob/bf993a1e5dc82216995138a97033ba4cc973333e/src/sentry/sentry_metrics/use_case_id_registry.py#L25) it was always meant to be a temporary mapping.

RaduW replied:
Not at the moment (unfortunately), there is still some places that I didn't touch (mostly because I don't quite know what they do) where we still use UseCaseKey and we convert to and from UseCaseID.

We'll have to have another go to remove it completely (maybe we could ask for help from the people that wrote slicing_rooter for example).

iambriccardo commented on src/sentry/snuba/metrics/fields/base.py:
```diff
@@ -196,11 +196,11 @@ def _get_entity_of_metric_mri(
 
     entity_keys_set: frozenset[EntityKey]
-    if use_case_id == UseCaseKey.PERFORMANCE:
+    if use_case_id is UseCaseID.TRANSACTIONS:
```
Why somewhere we are using `==` and here `is`? I would keep `==`, what do you think?

RaduW replied:
I guess different people changed it in different ways.
I generally use `==` with the exception of when I check against `None` where I always try to use `is`.

iambriccardo replied:
Exactly

RaduW replied:
`is UseCaseId.Transaction` is the correct usage according to PEP 8:

> Comparisons to singletons like None should always be done with is or is not, never the equality operators.

https://peps.python.org/pep-0008/#programming-recommendations

iambriccardo commented on src/sentry/sentry_metrics/utils.py:
```diff
@@ -15,9 +20,19 @@ class MetricIndexNotFound(InvalidParams):
+
+
 def reverse_resolve_tag_value(
-    use_case_id: UseCaseKey, org_id: int, index: Union[int, str, None], weak: bool = False
+    use_case_id: Union[UseCaseID, UseCaseKey],
+    org_id: int,
```
I agree with @RaduW on this. I would prefer to have all the downstream methods accept `UseCaseId` and perform coercions at top level if we don't want to also refactor there.

iambriccardo reviewed (commented):
Left some minor nits

RaduW commented on src/sentry/api/endpoints/organization_metrics.py:
```diff
@@ -21,17 +22,18 @@
 
 
-def get_use_case_id(request: Request) -> UseCaseKey:
+def get_use_case_id(request: Request) -> UseCaseID:
```
@ogi this is already happening, it is isolated in `string_to_use_case_id`:
that is used in `get_use_case_id`

obostjancic reviewed (approved):
LGTM 🚢
